File: javascript/AltiumClient/apiClient.js
const { URLSearchParams } = require("url");
const http = require("http");
const https = require("https");
const crypto = require("crypto");
const { spawn } = require("child_process");
const os = require("os");

a365Page = (title, message) => `
    <html>
    <head>
      <link href="https://fonts.googleapis.com/css?family=Montserrat:400,700" rel="stylesheet" type="text/css">
      <title>${title}</title>
      <style>
        html {
          height: 100%;
          background-image: linear-gradient(to right, #000b24, #001440);
        }
        body {
          color: #ffffff;
        }
        .center {
          width: 100%;
          position: absolute;
          left: 50%;
          top: 50%;
          transform: translate(-50%, -50%);
          text-align: center;
        }
        .title {
          font-family: Montserrat, sans-serif;
          font-weight: 400;
        }
        .normal {
          font-family: Montserrat, sans-serif;
          font-weight: 300;
        }
      </style>
    </head>
    <body>
      <div class="center">
        <h1 class="title">${title}</h1>
        <p class="normal">${message}.</p>
      </div>
    </body>
    </html>
    `;
const TOKEN_OPTIONS = {
  hostname: "auth.altium.com",
  path: "/connect/token",
  method: "POST",
  headers: {
    "Content-Type": "application/x-www-form-urlencoded",
  },
};

function launchBrowser(auth_request) {
  switch (os.platform()) {
    case "win32":
      return spawn("start", ['""', `"${auth_request}"`], { shell: true });
    case "darwin":
      return spawn("open", [`"${auth_request}"`], { shell: true });
    default:
      return spawn("xdg-open", [`"${auth_request}"`], { shell: true });
  }
}

function decodeJWT(jwt) {
  let json = JSON.parse(
    Buffer.from(
      jwt.split(".")[1].replace("-", "+").replace("_", "/"),
      "base64"
    ).toString("binary")
  );
  return json;
}

function getRequest(options, data) {
  return new Promise((resolve, reject) => {
    let req = https.request(options, (res) => {
      const contentType = res.headers["content-type"];

      let error;
      if (res.statusCode !== 200) {
        error = new Error(
          "Request Failed.\n" +
            `Status Code: ${res.statusCode} ${res.statusMessage}`
        );
      } else if (!/^application\/graphql-response|json/.test(contentType)) {
        error = new Error(
          "Invalid content-type.\n" +
            `Expected application/json but received ${contentType}`
        );
      }
      if (error) {
        console.error(error.message);
        // Consume response data to free up memory
        res.resume();
        return;
      }

      let rawData = "";
      res.setEncoding("utf8");
      res.on("data", (chunk) => (rawData += chunk));
      res.on("end", () => resolve(JSON.parse(rawData)));
    });
    req.on("error", (err) => reject(err));
    req.write(data);
    req.end();
  });
}

class AltiumClient {
  #exp;
  #accessToken;
  #pat;
  #scope;
  hostName = "usw2.dev-365.altium.com"; //"dev-365.altium.com/"; // "eur.365.altium.com";
  static scopes = {
    supply: "supply.domain",
    design: "openid profile email design.domain user.access offline_access",
  };

  /**
   * Client for the Altium 365 API to manage authorization and requests.
   * @param {string} pat - personal access token.
   * @param {string} [scope] - the resources required for authorization
   */

  constructor(pat, scope = AltiumClient.scopes.supply) {
    this.#pat = pat;
    this.#scope = scope;
  }

  set host(name) {
    this.hostName = name.replace(/^https:\/\//, "")
                        .replace(/\/graphql$/, "")
                        .replace(/\/gateway$/, "")
                        .replace(/\/napi$/, "")
                        .replace(/\/svc$/, "");
  }

  #refreshToken(token) {
    if ("refresh_token" in token) {
      const data = new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: token.refresh_token,
        client_id: this.#pat,
        scope: this.#scope,
      });

      return getRequest(TOKEN_OPTIONS, data.toString());
    }

    return new Promise(() => ""); // this.#getAccessToken(this.#id, this.#secret, this.#scope);
  }

  #checkTokenExp() {
    this.#exp =
      this.#exp ||
      this.#accessToken.then(
        (token) => decodeJWT(token)?.exp * 1000
      );

    return this.#exp.then((exp) => {
      if (exp < Date.now() + 300000) {
        //token is expired ... or will be in less than 5 minutes (300000 msec)
        this.#exp = undefined;
        this.#accessToken = this.#accessToken.then((token) =>
          this.#refreshToken(token)
        );
      }
      return this.#accessToken;
    });
  }

  /**
   * Make a request to the Altium 365 API
   * @param {string} gqlQuery - graphQL string containing the query/mutation.
   * @param {object} variables - key/value pairs for variables used in the gqlQuery.
   * @returns {object} - The Altium 365 API response
   */

  query(gqlQuery, path, variables) {
      this.#accessToken =
          this.#accessToken ||
          Promise.resolve(this.#pat);
      
    return this.#checkTokenExp().then((token) => {
      const options = {
        hostname: this.hostName,
        path: path,
        method: "POST",
        headers: {
          Authorization: "Bearer " + token,
          "Content-Type": "application/json",
        },
      };
      const data = {
        query: gqlQuery,
        variables: variables,
      };

      return getRequest(options, JSON.stringify(data));
    });
  }

  /**
   * Iterable for a graphQL Type implementing a node interface.
   * NB: the query must include a variable to set the the cursor and the pageInfo field on the Type.
   * @async
   * @generator
   * @param {string} pageKey - graphQL variable name for setting the cursor: desProjects(after: $pageKey).
   * @param {function} pageSelect - return from response the type with node interface: (data) => data.desProjects
   * @yields {object} - a page of the graphQL Type implementing a node interface
   */

  async *pageGen(gqlQuery, path, gqlVariables, pageKey, pageSelect) {
    let pageInfo = { hasNextPage: true };
    while (pageInfo.hasNextPage) {
      const response = await this.query(gqlQuery, path, gqlVariables);

      pageInfo = pageSelect(response.data).pageInfo;
      gqlVariables[pageKey] = pageInfo.endCursor;

      yield pageSelect(response.data).nodes;
    }
  }
}

module.exports = { AltiumClient };

File: python/AltiumClient/apiClient.py
"""Resources for making A365 requests."""
import requests, re
import base64
import json
import time
from typing import Callable, Dict, Iterator

import jwt
from a365Token import get_access_token_using_refresh_token

A365_URL =  "https://usw2.dev-365.altium.com/napi/gateway/graphql" # "https://eur.365.altium.com/napi/gateway/graphql"

def decodeJWT(token):
    return json.loads(
        (base64.urlsafe_b64decode(token.split(".")[1] + "==")).decode("utf-8")
    )

class AltiumClient:
    def __init__(self, id, secret, refresh_token, pat, scopes = ['supply.domain']) -> None:
        #scopes = ['supply.domain', 'design.domain', 'user.access', 'offline_access']
        self.id = id
        self.secret = secret
        self.refresh_token = refresh_token
        self.pat = pat
        self.scopes = scopes
        self.api_url = A365_URL
        self.s = requests.session()
        self.s.keep_alive = False

        if pat is not None:
            self.token = pat
        elif all(v is not None for v in (id, secret, refresh_token)):
            self.token = get_access_token_using_refresh_token(id, secret, refresh_token, scopes)
            #self.s.headers.update({"token": self.token.get('access_token')})
            #self.s.headers.update({"Authorization": f"Bearer {self.token.get('access_token')}"})
            self.exp = decodeJWT(self.token.get('access_token')).get('exp')

    def check_exp(self):
        if (self.exp < time.time() + 300):
            self.token = get_access_token_using_refresh_token(self.id, self.secret, self.refresh_token, self.scopes)
            #self.s.headers.update({"token": self.token.get('access_token')})
            #self.s.headers.update({"Authorization": f"Bearer {self.token.get('access_token')}"})
            self.exp = decodeJWT(self.token.get('access_token')).get('exp')
            
    def token_workspace_scope_match(self, workspace_guid):
        # If self.token is a string (e.g. PAT case), convert it to a dictionary (scope in that case will be an array, not str)
        if isinstance(self.token, str):
            token_data = jwt.decode(self.token, options={"verify_signature": False})   
        else:
            token_data = self.token
        
        a365_ws_prefix = "a365:workspace:"
        scopes = token_data.get('scope', [])
        scopes = scopes.split() if isinstance(scopes, str) else list(scopes)
        scopes = [s.lower().removeprefix(a365_ws_prefix) for s in scopes]
        return workspace_guid.lower() in scopes
        

    def get_query(self, query: str, variables: Dict = {}) -> dict:
        """Return A365 response for the query."""
        try:
            if self.pat is not None:
                r = self.s.post(
                    self.api_url,
                    json={"query": query, "variables": variables},
                    headers={"Authorization": f"Bearer {self.pat}"}
                )
            else:
                self.check_exp()      
                temp = self.token.get('access_token')
                r = self.s.post(
                    self.api_url,
                    json={"query": query, "variables": variables},
                    headers={"Authorization": f"Bearer {temp}"}
                )

        except Exception as e:
            print(e)
            raise Exception("Error while getting A365 response")

        response = r.json()
        if ("errors" in response):
            for error in response["errors"]: print(error["message"])
            raise SystemExit

        return response["data"]

    class Node:
        def __init__(self, client, query: str, variables: Dict, f: Callable) -> None:
            self.client = client
            self.query = query
            self.variables = variables
            self.f = f
            self.name = re.search("after[\s]*:[\s]*\$([\w]*)", query).group(1)

        def __iter__(self) -> Iterator:
            self.pageInfo = {"hasNextPage": True}
            return self
 
        def __next__(self):
            if (not self.pageInfo["hasNextPage"]): raise StopIteration

            data = self.client.get_query(self.query, self.variables)

            self.pageInfo = self.f(data)["pageInfo"]
            self.variables[self.name] = self.pageInfo["endCursor"]
            return self.f(data)["nodes"]

    def NodeIter(self, query: str, variables: dict, f: Callable) -> Iterator:
        return AltiumClient.Node(self, query, variables, f)
